TPA modes some refactoring

limonspb opened the pull request:
This refactoring would allow to easier modify/improve/understand TPA code behavior.

## Old description (not relevant anymore)
([read this comment](https://github.com/betaflight/betaflight/pull/13804#issuecomment-2292617666))

Not sure how to make it nicer. Current implementation downside: the I-term will be logged before TPA factor.
Upside: it's simple to follow, and works :)

I-term for wings works like a "dynamic servo trim", always trying to keep craft on rails.
However as wing speed grows, wing needs less I-term (less "default" servo angles) to keep it straight.
And when speed drops, we have to go back to the old bigger value, **without waiting for I-term to accumulate naturally.**

Please, let me know if anyone has an idea how to make it nicer. Can't really do: `pidData[axis].I *= tpaFactorKi`, because the next PID loop it's gonna do it again, making I-term zero pretty quick. (because I-term has accumulative effect, so it depends on previous values).

@KarateBrot  @ledvinap  @ctzsnooze @nerdCopter

ledvinap commented:
@limonspb : IMO the problem is that `pidData[axis].I` is used both as `I` integrator and actual `I` value. Splitting it into two variables will help a bit (IMO SPA has similar problem).

But current PID code is .. interesting. Touching it will have consequences

limonspb commented:
> @limonspb : IMO the problem is that `pidData[axis].I` is used both as `I` integrator and actual `I` value. Splitting it into two variables will help a bit (IMO SPA has similar problem).
> 
> But current PID code is .. interesting. Touching it will have consequences

Yeah I was trying to unwrap I term, but didn't work out nicely...

Maybe just this, but also add I-term * tpaFactor to TPA debug?

nerdCopter commented:
commented something in discord :wink: 

also, maybe the `tpaFactorKi` should be applied in the `// -----calculate I component` section.

ledvinap commented:
> also, maybe the tpaFactorKi should be applied in the // -----calculate I component section (i'm not sure, but seems so).

That's the idea. But it is impossible(or impractical) with single `.I` - you don't want to decrease actual I-term accumulator value, just I-term influence on pid

nerdCopter commented:
> > also, maybe the tpaFactorKi should be applied in the // -----calculate I component section (i'm not sure, but seems so).
> 
> That's the idea. But it is impossible(or impractical) with single `.I` - you don't want to decrease actual I-term accumulator value, just I-term influence on pid

Makes sense. i was thinking `// -----calculate I component` because the P and the D are in their respective sections.

Could they all safely be applied (moved) directly before the `const Pisum`?  (This is how another firmware situates it.)

limonspb commented:
> > > also, maybe the tpaFactorKi should be applied in the // -----calculate I component section (i'm not sure, but seems so).
> > 
> > 
> > That's the idea. But it is impossible(or impractical) with single `.I` - you don't want to decrease actual I-term accumulator value, just I-term influence on pid
> 
> Makes sense. i was thinking `// -----calculate I component` because the tpaFactor for P and for the D are in their respective sections.
> 
> Could they all safely be applied (moved) directly before the `const Pisum`? (This is how another firmware situates it.)

Do you mean just move tpaFactorKi definition to right before pidsum calculation?

I was just trying to avoid extra `#ifdef USE_TPA_MODE`

ledvinap commented:
@limonspb : You can't refactor I TPA easily - so instead refactor P/D TPA to match I.

limonspb commented:
> @limonspb : You can't refactor I TPA easily - so instead refactor P/D TPA to match I.

Do you mean just moving `tpaFactorK*` calculations for right before `const float pidSum =`?
Could you please elaborate.

limonspb commented on src/main/flight/pid.c:
```diff
@@ -1225,7 +1227,7 @@ void FAST_CODE pidController(const pidProfile_t *pidProfile, timeUs_t currentTim
 
         // calculating the PID sum
-        const float pidSum = pidData[axis].P + pidData[axis].I + pidData[axis].D + pidData[axis].F + pidData[axis].S;
+        const float pidSum = pidData[axis].P + pidData[axis].I * tpaFactorKi + pidData[axis].D + pidData[axis].F + pidData[axis].S;
```
In this case for blackbox none of the terms will be recorded after TPA. Is that ok?

ledvinap replied:
I didn't think about that.
For blackbox, I will be problematic in both cases - sometimes you need I accumulator, sometimes I value.

For quick fix, adding `blackbox_I` to `pidData` is possible.
And I'd move `pidData[axis].P *= tpaFactorKp, `pidData[axis].D *= tpaFactorKd` here anyway, to better separate TPA from PID  (`const tpaFactorK<x> = 1.0f` will be optimized away, so there would be no penalty without TPA)

limonspb commented:
@ledvinap i did some update, tried to abstract with `getTpaFactors` function. Added `pidData[i].I_afterTpa` that is also used for bbx. And added tpa factors for FF and S-term. They are always 1.0 for now, but in future im also planning to add modes like
TPA_MODE_PIDFF, TPA_MODE_PIDFFS etc.

ledvinap commented on src/main/flight/pid.c:
```diff
@@ -1220,14 +1250,17 @@ void FAST_CODE pidController(const pidProfile_t *pidProfile, timeUs_t currentTim
             // yaw has a tendency to windup. Otherwise limit yaw iterm accumulation.
             const int launchControlYawItermLimit = (pidRuntime.launchControlMode == LAUNCH_CONTROL_MODE_FULL) ? LAUNCH_CONTROL_YAW_ITERM_LIMIT : 0;
             pidData[FD_YAW].I = constrainf(pidData[FD_YAW].I, -launchControlYawItermLimit, launchControlYawItermLimit);
+            pidData[FD_YAW].I_afterTpa = pidData[FD_YAW].I;
 
             // for pitch-only mode we disable everything except pitch P/I
             if (pidRuntime.launchControlMode == LAUNCH_CONTROL_MODE_PITCHONLY) {
                 pidData[FD_ROLL].P = 0;
                 pidData[FD_ROLL].I = 0;
+                pidData[FD_ROLL].I_afterTpa = 0;
                 pidData[FD_YAW].P = 0;
                 // don't let I go negative (pitch backwards) as front motors are limited in the mixer
                 pidData[FD_PITCH].I = MAX(0.0f, pidData[FD_PITCH].I);
+                pidData[FD_PITCH].I_afterTpa = pidData[FD_PITCH].I;
```
Poor code structure manifests in this way...

limonspb replied:
I don't see how I can make this PR the other way...

ledvinap replied:
Sorry, just a comment. It is problem with PID code, not this PR ...

limonspb replied:
> Sorry, just a comment. It is problem with PID code, not this PR ...

could you please trigger a rebuild for the PR? I don't have privileges. AT32 targets failed, i don't see why.

ledvinap replied:
It seems to fail on resource limit. I did restart it once and it failed again. But I don't understand github CI

limonspb replied:
Seems good now! thanks

nerdCopter replied:
i triggered it again and enabled debugging, but it succeeded, so nothing to debug i suppose.

limonspb commented:
i've tested this PR with wing flights. Tpa mode PID doesn't actually do what i was thinking.
On a straight line flight the I-term mostly just sets servos to a proper flat position, that can be different from "zero" position.
And if we start changing I-term depending on throttle, the control surfaces will go off from the flat position.

I have removed PID from the PR, but i think the refactoring here is good, allowing easier to add, modify or understand tpa modes. @ledvinap

ledvinap commented:
@limonspb : With only one TPA factor, it's a bit too complicated. But other 'factor' stuff may follow in future, making PID a bit cleaner.

limonspb commented:
> @limonspb : With only one TPA factor, it's a bit too complicated. But other 'factor' stuff may follow in future, making PID a bit cleaner.

Agree, that could be too complicated for now.

For S-term multiplier, i guess, i can just do the math in getSterm function we already have.

More complicated will be to handle YAW TPA. Because YAW can be either rudder, or differential thrust. (or even both, but i'm not thinking about it yet). So when YAW is done with differential thrust, we need to fall back to the a classic TPA curve, and proabaly to a classic TPA argument, which is just throttle.

ledvinap commented:
side note: did you test differential thrust? in my very limited experience, it is very ineffective at yawing plane..

limonspb commented:
> side note: did you test differential thrust? in my very limited experience, it is very ineffective at yawing plane..

i been enjoying it a lot actually.
Indeed you can't use it as a real YAW thing, but as a yaw stabilizer it's doing very well.
Basically i set:
1. yaw s-term to zero
2. yaw I-term zero
3. yaw rates "almost zero", like 50 degrees/sec max.

So i left it with just P and D gains, and it eliminated the "yaw wondering".
It also helps to keep nose horizontal when doing knife edging (when roll 90 degrees).
And of course it's very useful in 3D hover, can use yaw stick almost like on the drone.

At high speeds can't really yaw with the stick, because vertical stabilizers are too strong.
But at lower speeds i can use some yaw. Just have to be careful not to use it at speeds close to stall, it will spin the plane like crazy :)

Just differential thrust yaw will need a different TPA...
